reflect.Set: value of type interface {} is not assignable to type interp.valueInterface

yaegi version: 0.8.13

The following program `sample.go` triggers a panic:

```go
package main

import "fmt"

// Filter is a filter interface
type Filter interface {
	Bounds(srcBounds string) (dstBounds string)
}

// GIFT type
type GIFT struct {
	Filters []Filter
}

// New creates a new filter list and initializes it with the given slice of filters.
func New(filters ...Filter) *GIFT {
	return &GIFT{
		Filters: filters,
	}
}

// Bounds calculates the appropriate bounds for the result image after applying all the added filters.
func (g *GIFT) Bounds(srcBounds string) (dstBounds string) {
	dstBounds = srcBounds
	for _, f := range g.Filters {
		dstBounds = f.Bounds(dstBounds)
	}
	return dstBounds
}

type resizeFilter struct{}

func (p *resizeFilter) Bounds(srcBounds string) (dstBounds string) {
	return srcBounds
}

func main() {
	var filters []Filter
	bounds := "foo"
	g := New(filters...)
	fmt.Println(g.Bounds(bounds))
}
```

Expected result:
```console
$ go run ./sample.go
foo
```

Got:
```console
$ yaegi ./sample.go
./sample.go:24:2: panic
./sample.go:38:6: panic
reflect.Set: value of type interface {} is not assignable to type interp.valueInterface
goroutine 1 [running]:
runtime/debug.Stack(0x1, 0xc000382800, 0x40)
        /usr/lib/go-1.14/src/runtime/debug/stack.go:24 +0x9d
github.com/containous/yaegi/interp.(*Interpreter).Eval.func1(0xc0002a9dd8)
        /home/me/src/softwares/go/src/github.com/containous/yaegi/interp/interp.go:333 +0xbc
panic(0xd2ed00, 0xc0000ab3e0)
        /usr/lib/go-1.14/src/runtime/panic.go:969 +0x166
github.com/containous/yaegi/interp.runCfg.func1(0xc0004300a0, 0xc000445800)
        /home/me/src/softwares/go/src/github.com/containous/yaegi/interp/run.go:119 +0x20c
panic(0xd2ed00, 0xc0000ab3e0)
        /usr/lib/go-1.14/src/runtime/panic.go:969 +0x166
github.com/containous/yaegi/interp.runCfg.func1(0xc000430280, 0xc000442400)
        /home/me/src/softwares/go/src/github.com/containous/yaegi/interp/run.go:119 +0x20c
panic(0xd2ed00, 0xc0000ab3e0)
        /usr/lib/go-1.14/src/runtime/panic.go:969 +0x166
reflect.Value.assignTo(0xd5f100, 0xc0000ab3b0, 0x194, 0xe6b731, 0xb, 0xdb6180, 0x0, 0xc0000ab3c0, 0xe6b731, 0x17)
        /usr/lib/go-1.14/src/reflect/value.go:2403 +0x426
reflect.Value.Set(0xdb6180, 0xc0000afa20, 0x199, 0xd5f100, 0xc0000ab3b0, 0x194)
        /usr/lib/go-1.14/src/reflect/value.go:1532 +0xbd
github.com/containous/yaegi/interp._range.func1(0xc000430280, 0xc0003b5140)
        /home/me/src/softwares/go/src/github.com/containous/yaegi/interp/run.go:2177 +0x1cd
github.com/containous/yaegi/interp.runCfg(0xc000442400, 0xc000430280)
        /home/me/src/softwares/go/src/github.com/containous/yaegi/interp/run.go:125 +0x6d
github.com/containous/yaegi/interp.call.func7(0xc0004300a0, 0xc00045a070)
        /home/me/src/softwares/go/src/github.com/containous/yaegi/interp/run.go:1023 +0xc42
github.com/containous/yaegi/interp.runCfg(0xc000445800, 0xc0004300a0)
        /home/me/src/softwares/go/src/github.com/containous/yaegi/interp/run.go:125 +0x6d
github.com/containous/yaegi/interp.(*Interpreter).run(0xc0000cc1a0, 0xc000445000, 0xc000430000)
        /home/me/src/softwares/go/src/github.com/containous/yaegi/interp/run.go:103 +0x247
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc0000cc1a0, 0xc00043a000, 0x351, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/me/src/softwares/go/src/github.com/containous/yaegi/interp/interp.go:416 +0x5f7
main.main()
        /home/me/src/softwares/go/src/github.com/containous/yaegi/cmd/yaegi/yaegi.go:166 +0x6e2
```

--- a/_test/issue-784.go
+++ b/_test/issue-784.go
@@ -28,12 +28,6 @@ func (g *GIFT) Bounds(srcBounds string) (dstBounds string) {
 	return dstBounds
 }
 
-type resizeFilter struct{}
-
-func (p *resizeFilter) Bounds(srcBounds string) (dstBounds string) {
-	return srcBounds
-}
-
 func main() {
 	var filters []Filter
 	bounds := "foo"

--- a/_test/issue-784.go
+++ b/_test/issue-784.go
@@ -1,0 +1,45 @@
+package main
+
+import "fmt"
+
+// Filter is a filter interface
+type Filter interface {
+	Bounds(srcBounds string) (dstBounds string)
+}
+
+// GIFT type
+type GIFT struct {
+	Filters []Filter
+}
+
+// New creates a new filter list and initializes it with the given slice of filters.
+func New(filters ...Filter) *GIFT {
+	return &GIFT{
+		Filters: filters,
+	}
+}
+
+// Bounds calculates the appropriate bounds for the result image after applying all the added filters.
+func (g *GIFT) Bounds(srcBounds string) (dstBounds string) {
+	dstBounds = srcBounds
+	for _, f := range g.Filters {
+		dstBounds = f.Bounds(dstBounds)
+	}
+	return dstBounds
+}
+
+type resizeFilter struct{}
+
+func (p *resizeFilter) Bounds(srcBounds string) (dstBounds string) {
+	return srcBounds
+}
+
+func main() {
+	var filters []Filter
+	bounds := "foo"
+	g := New(filters...)
+	fmt.Println(g.Bounds(bounds))
+}
+
+// Output:
+// foo

--- a/interp/value.go
+++ b/interp/value.go
@@ -156,34 +156,35 @@ func genValueArray(n *node) func(*frame) reflect.Value {
 func genValueRangeArray(n *node) func(*frame) reflect.Value {
 	value := genValue(n)
 
-	// dereference array pointer, to support array operations on array pointer
-	if n.typ.TypeOf().Kind() == reflect.Ptr {
+	switch {
+	case n.typ.TypeOf().Kind() == reflect.Ptr:
+		// dereference array pointer, to support array operations on array pointer
 		return func(f *frame) reflect.Value {
 			return value(f).Elem()
 		}
-	}
-	if n.typ.val.cat == interfaceT {
+	case n.typ.val.cat == interfaceT:
 		return func(f *frame) reflect.Value {
 			val := value(f)
 			v := []valueInterface{}
 			for i := 0; i < val.Len(); i++ {
-				if av, ok := val.Index(i).Interface().([]valueInterface); ok {
+				switch av := val.Index(i).Interface().(type) {
+				case []valueInterface:
 					v = append(v, av...)
-				} else if av, ok := val.Index(i).Interface().(valueInterface); ok {
+				case valueInterface:
 					v = append(v, av)
-				} else {
+				default:
 					panic(n.cfgErrorf("invalid type %v", val.Index(i).Type()))
 				}
 			}
 			return reflect.ValueOf(v)
 		}
-	}
-
-	return func(f *frame) reflect.Value {
-		// This is necessary to prevent changes in the returned
-		// reflect.Value being reflected back to the value used
-		// for the range expression.
-		return reflect.ValueOf(value(f).Interface())
+	default:
+		return func(f *frame) reflect.Value {
+			// This is necessary to prevent changes in the returned
+			// reflect.Value being reflected back to the value used
+			// for the range expression.
+			return reflect.ValueOf(value(f).Interface())
+		}
 	}
 }
 

--- a/interp/value.go
+++ b/interp/value.go
@@ -155,12 +155,30 @@ func genValueArray(n *node) func(*frame) reflect.Value {
 
 func genValueRangeArray(n *node) func(*frame) reflect.Value {
 	value := genValue(n)
+
 	// dereference array pointer, to support array operations on array pointer
 	if n.typ.TypeOf().Kind() == reflect.Ptr {
 		return func(f *frame) reflect.Value {
 			return value(f).Elem()
 		}
 	}
+	if n.typ.val.cat == interfaceT {
+		return func(f *frame) reflect.Value {
+			val := value(f)
+			v := []valueInterface{}
+			for i := 0; i < val.Len(); i++ {
+				if av, ok := val.Index(i).Interface().([]valueInterface); ok {
+					v = append(v, av...)
+				} else if av, ok := val.Index(i).Interface().(valueInterface); ok {
+					v = append(v, av)
+				} else {
+					panic(n.cfgErrorf("invalid type %v", val.Index(i).Type()))
+				}
+			}
+			return reflect.ValueOf(v)
+		}
+	}
+
 	return func(f *frame) reflect.Value {
 		// This is necessary to prevent changes in the returned
 		// reflect.Value being reflected back to the value used

